### PR TITLE
Implement the Key Derivation API

### LINKED ISSUE #63: Support Key derivation API
https://doc.libsodium.org/key_derivation

### PR DESCRIPTION
**Pull request by Kleidukos**
Fixes #63

### CODE REVIEW

File: libsodium-bindings/src/LibSodium/Bindings/KeyDerivation.hs [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+    -- ^ Pointer that will hold the sub-key.
+    -> CSize
+    -- ^ Length of the sub-key.
+    -> Word64
+    -- ^ Identifier of the sub-key. Must not be reused for another sub-key.
+    -> Ptr CChar
+    -- ^ Pointer to the context, of size 'cryptoKDFContextBytes'.
+    -> Ptr CUChar
+    -- ^ Pointer to the master key, which will be of length 'cryptoKDFKeyBytes'.
+    -> IO CInt
```
Comment (kozross) [inline libsodium-bindings/src/LibSodium/Bindings/KeyDerivation.hs:49]: What does the return value mean?

File: libsodium-bindings/src/LibSodium/Bindings/KeyDerivation.hs [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+
+-- == Constants
+
+-- | The algorithm used by the key derivation algorithm, blake2b.
+--
+-- @since 0.0.1.0
+cryptoKDFPrimitive :: String
+cryptoKDFPrimitive = "blake2b"
```
Comment (kozross) [inline libsodium-bindings/src/LibSodium/Bindings/KeyDerivation.hs:57]: Is this something you're meant to pass into C? If so, `String` won't work, you need a C string.

**Review by kozross — CHANGES_REQUESTED** [review]

File: libsodium-bindings/src/LibSodium/Bindings/KeyDerivation.hs (hunk 2 above)
Comment (Kleidukos) [reply libsodium-bindings/src/LibSodium/Bindings/KeyDerivation.hs:57]: No, it's a fixed string on the C side but the C compiler is giving me shit about it.

```c

#define crypto_kdf_PRIMITIVE "blake2b"
SODIUM_EXPORT
const char *crypto_kdf_primitive(void)
            __attribute__ ((warn_unused_result));
```

I tried a lot of export types, and it was never quite right, so I gave up and just re-used the string.

File: libsodium-bindings/src/LibSodium/Bindings/KeyDerivation.hs (hunk 1 above)
Comment (Kleidukos) [reply libsodium-bindings/src/LibSodium/Bindings/KeyDerivation.hs:49]: Ah good catch

File: libsodium-bindings/src/LibSodium/Bindings/KeyDerivation.hs (hunk 2 above)
Comment (kozross) [reply libsodium-bindings/src/LibSodium/Bindings/KeyDerivation.hs:58]: What exact error are you getting? A 'fixed string on the C side' is _nothing whatsoever_ like a Haskell `String`, so whatever the cause, your solution won't work.

Comment (kozross) [reply libsodium-bindings/src/LibSodium/Bindings/KeyDerivation.hs:58]: Does `Ptr CChar` not work?

**Review by kozross — APPROVED** [review]

**Pull request merged**